### ISSUE TITLE
Mocking structs from other crates

### ISSUE THREAD

**Issue by hardskulls** [asomers/mockall#404]
I need to quicly create structs from other crate(s) (just like with `::default()` or `::new()` methods).
`#[double]` macro doesn't work in this case.
Is there a way to do it?

**Comment by asomers** [comment]
Not automatically.  `#[double]` doesn't work because the other crate doesn't publish any mocks.  But you can still use `mock!`.  That's what you'll have to do.

**Comment by hardskulls** [comment]
```rust
mock! {
    use some_crate::module::some_struct;
}
```
Something like that?

**Comment by asomers** [comment]
Nope.  You basically have to repeat the trait definition within the `mock!` block.

**Comment by asomers** [comment]
Oh, then `mock!` the other crate's struct definition then.  Do it just like you would use `mock!` for a struct from your own crate.

**Comment by hardskulls** [comment]
Rewrite the whole struct definition? That's exactly what I wanted to avoid :(
What about traits implementations from the crate the struct is defined in? Do I have to add them manually in `mock!`?

The problem to begin with is that I have a crate that contains functionality based on other crate, but to test my functions I need to create structs from the crate that I'm using.
The structs don't have `new` methods on them.
While yes, I can create them manually, creating a pretty nested struct bothers me a bit.
And I can't use `#[derive(Default)]` here because these structs are not from my crate,.

**Comment by asomers** [comment]
Yes.  Unless the other crate happens to export mocks, you'll have to redefine its structs using `mock!`.  Mockall simply has no way to know what extern crates definitions are.  And if those structs have trait implementations, then you'll have to include those in your `mock!`.

**Comment by hardskulls** [comment]
That's unfortunate :(
Thanks for the help anyway.
Is it limitation of rust itself or just `mockall`?

**Comment by asomers** [comment]
That's a Rust limitation.  Proc macros get evaluated much too early to know about the contents of other crates.
